drivers/vdpa/virtio/virtio_vdpa.c

```diff
@@ -552,100 +552,145 @@ virtio_vdpa_vring_state_set(int vid, int vq_idx, int state)
 }
 
 static int
-virtio_vdpa_dma_unmap(struct virtio_vdpa_priv *priv)
+virtio_vdpa_dev_cleanup(int vid)
 {
 	uint32_t i;
 	int ret;
-	struct rte_vhost_memory *mem = NULL;
-	int vfio_container_fd;
+	struct rte_vhost_memory *mem;
+	struct rte_vdpa_device *vdev = rte_vhost_get_vdpa_device(vid);
+	struct virtio_vdpa_priv *priv =
+		virtio_vdpa_find_priv_resource_by_vdev(vdev);
 
-	ret = rte_vhost_get_mem_table(priv->vid, &mem);
-	if (ret < 0) {
-		DRV_LOG(ERR, "%s failed to get VM memory layout ret:%d",
-					priv->vdev->device->name, ret);
-		goto exit;
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s", vdev->device->name);
+		return -ENODEV;
 	}
 
-	vfio_container_fd = priv->vfio_container_fd;
+	mem = priv->mem;
+	if (mem == NULL) {
+		DRV_LOG(INFO, "No mem is registered: %s", vdev->device->name);
+		return 0;
+	}
 
 	for (i = 0; i < mem->nregions; i++) {
 		struct rte_vhost_mem_region *reg;
 
 		reg = &mem->regions[i];
 		DRV_LOG(INFO, "%s, region %u: HVA 0x%" PRIx64 ", "
 			"GPA 0x%" PRIx64 ", size 0x%" PRIx64 ".",
-			"DMA unmap", i,
+			"DMA clean up unmap", i,
 			reg->host_user_addr, reg->guest_phys_addr, reg->size);
 
-		ret = rte_vfio_container_dma_unmap(vfio_container_fd,
+		ret = rte_vfio_container_dma_unmap(priv->vfio_container_fd,
 			reg->host_user_addr, reg->guest_phys_addr,
 			reg->size);
 		if (ret < 0) {
-			DRV_LOG(ERR, "%s DMA unmap failed ret:%d",
+			DRV_LOG(ERR, "%s DMA clean up failed ret:%d",
 						priv->vdev->device->name, ret);
-			goto exit;
+			free(mem);
+			priv->mem = NULL;
+			return ret;
 		}
 	}
 
-exit:
 	free(mem);
-	return ret;
+	priv->mem = NULL;
+	return 0;
 }
+static inline bool
+virtio_vdpa_find_mem_reg(struct rte_vhost_mem_region *key,struct rte_vhost_memory *mem)
```

Comment:
make key and mem const pointer as this search routine is not going to modify anything.